### PR TITLE
[Messenger] ReceiverInterface::handle() to get() & Worker with prioritized transports

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | Helps with #30699
| License       | MIT
| Doc PR        | TODO

Highlights:

* `messenger:consume` can now consume messages from multiple transports with priority ❗️ 

```
bin/console messenger:consume amqp_high amqp_medium amqp_low
```

* How long you want to sleep before checking more messages is now an option to `messenger:consume`
* `ReceiverInterface::receive()` is replaced with `ReceiverInterface::get()`
* Logic for looping & sleeping is moved into `Worker`

### CODE REVIEW

File: src/Symfony/Component/Messenger/Worker.php [hunk 1]
```diff
@@ -60,21 +61,33 @@ public function __construct(ReceiverInterface $receiver, MessageBusInterface $bu
...
+        while (false === $this->shouldStop) {
+            $envelope = $this->receiver->get();
+
             if (null === $envelope) {
-                if (\function_exists('pcntl_signal_dispatch')) {
-                    pcntl_signal_dispatch();
-                }
+                $handled(null);
+
+                usleep(1000000);
```
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Worker.php]: Two changes here:

1) I changed from .2 seconds to 1 second. Waiting only 200ms between when there is not a message is quite rapid. For reference, in Laravel, this defaults to 3 and is configurable on the command/worker.

**Comment by weaverryan** [comment]
This is ready!

tl;dr In recent changes, the `SenderInterface` is not really the layer that's handling the "looping and waiting for messages" layer -  the `Worker` is doing that. This completes that transformation. It will also allow us to create a `Worker` that listens to multiple queues at once, in a priority order (#30699).

File: src/Symfony/Component/Messenger/CHANGELOG.md [hunk 2]
```diff
@@ -10,7 +10,8 @@ CHANGELOG
    the transport. See `ConsumeMessagesCommand`.
  * The optional `$busNames` constructor argument of the class `ConsumeMessagesCommand` was removed.
  * [BC BREAK] 2 new methods were added to `ReceiverInterface`:
-   `ack()` and `reject()`.
+   `ack()` and `reject()`. Also `receive()` was changed to `get()`
+   and `stop()` was removed.
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/CHANGELOG.md]: `receive` was not really changed to `get`, was it? It's more like `received` was removed and `get` added, right?

**Comment by sroze** [comment]
The reasoning of having `handle($handler)` was for the transports to be able to catch the exceptions from the handlers and deal with the errors (i.e. ack, nack, etc...). Since we moved this to the worker, I'm happy to move back to get one message. Though, why wouldn't we return an `iterable` instead to match the behaviour of prefetching (i.e. some transports would have loaded X messages, not just one)?

File: src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php [hunk 3]
```diff
@@ -25,21 +25,15 @@ interface ReceiverInterface
...
+     *
+     * @return Envelope[]
      */
-    public function receive(callable $handler): void;
-
-    /**
-     * Stop receiving some messages.
-     */
-    public function stop(): void;
+    public function get(): iterable;
```
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php]: The iterable, like always, was kind of a pain, compared to just returning one. But, it *is* more flexible and end-users won't touch this method much anyways.

File: src/Symfony/Component/Messenger/Worker.php [hunk 4]
```diff
@@ -32,112 +32,141 @@
...
-                    pcntl_signal_dispatch();
+                    $this->handleMessage($envelope, $receiver, $receiverName, $this->retryStrategies[$receiverName] ?? null);
+                    $handled($envelope);
                 }
 
-                return;
+                // after handling a single receiver, quit and start the loop again
+                // this should prevent multiple lower priority receivers from
+                // blocking too long before the higher priority are checked
+                if ($envelopeHandled) {
```
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Worker.php]: Important decision: once we get at least one message from one receiver, we immediately go back to the beginning of the loop. The idea is that if you have `high`, `medium`, `low` transports, if there are none on `high`, but messages on `medium` and `low`, the medium ones *will* be processed, but then it will reset and look once again for new `high` messages, instead of first processing any `low` messages.

**Comment by weaverryan** [comment]
This is ready for review! Description updated.

Better viewed (especially the `Worker` with `?w=1`). The 3 `StopWhen***` classes are not new - but were moved & changed enough (they decorate `WorkerInterface` instead of `ReceiverInerface` so they look new.

File: src/Symfony/Component/Messenger/Worker.php (hunk 4 above)
Comment (Nyholm) [reply src/Symfony/Component/Messenger/Worker.php]: There is a flaw here (I think). 

T0: `medium` got one million messages. 
T1: We check `high`, no messages. So we moving on 
T2: We starting to consume the messages in `medium`
T5: We add one messages to `high`
T100: We still handing messages in `medium`...
T101: One million more messages are added in `medium`
T1000: We still handing messages in `medium`...
T∞: We handle the message in `high`.

Comment (Nyholm) [reply src/Symfony/Component/Messenger/Worker.php]: We maybe should introduce a limit, say we only handle X messages from `medium` before rechecking `high`

**Review by Nyholm — CHANGES_REQUESTED** [review]
I really like this. 

I am not a fan of priority in a message system, but this is priority of transports. That is a different thing. Im all 👍 
(after my comments are addressed/answered)

File: src/Symfony/Component/Messenger/Worker.php (hunk 4 above)
Comment (weaverryan) [reply src/Symfony/Component/Messenger/Worker.php]: Indeed. So, this "may" not be a problem ;). And it might be an issue of documentation on the `ReceiverInterface::get()` method. The idea is that you might return a "few" messages, but never all of them. This was motivated by Google PubSub, where you *can* just ask for one message, but I guess you normally "pull" more than 1: http://googleapis.github.io/google-cloud-php/#/docs/cloud-pubsub/v1.9.0/pubsub/subscription?method=pull

The intention is more to return just the one immediate message... or "few" immediate messages. Another example where this will come up is when we support multiple queues for one AMQP transport. In `ReceiverInterface::send()`, the cleanest implementation is to loop over  ALL the queues each time and return the "one or zero" messages from each. But that means `get()` would still max out at returning a number of messages equal to your number of queues, which will not be a HUGE number. 

I'm going to attempt to solve this with some PHPdoc.

New addition to the phpdoc above `ReceiverInterface::get()`

```php
/**
 * ... 
 * While this method could return an unlimited number of messages,
 * the intention is that it returns only one, or a "small number"
 * of messages each time. This gives the user more flexibility:
 * they can finish processing the one (or "small number") of messages
 * from this receiver and move on to check other receivers for messages.
 * If a this method returns too many messages, it could cause a
 * blocking effect where handling the messages received from one
 * call to get() takes a long time, blocking other receivers from
 * being called.
 * ...
```

Comment (weaverryan) [reply src/Symfony/Component/Messenger/Worker.php]: The alternative, as you suggested, is to (in `Worker`) only process some "limit" of messages from each Receiver before stopping. That should work fine in a system that properly supports `ack()`... which is definitely the normal use-case. I'm just not sure if there will be some case where some weird transport considers the messages we received as, sort of, auto-acked, and so if we just "drop" them in loop above this code, they will never be redelivered.

**Review by Nyholm — APPROVED** [review]
Good. I’m happy with the updated doc block on ReceiverInterface::get(). 

Thank you

**Review by fabpot — APPROVED** [review]
minor typos

**Comment by fabpot** [comment]
Thank you @weaverryan.

**Comment by fabpot** [comment]
typos fixed in f4176b0cf6

File: src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php [hunk 5]
```diff
@@ -25,21 +25,25 @@ interface ReceiverInterface
     /**
      * Receive some messages to the given handler.
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php]: This doc is outdated: There is no "given handler"

**Comment by Tobion** [comment]
Just thinking out loud for reference: The prioritized transports can be used to prioritize messages by publishing messages to different transports and consuming them in order. But it's a different way than rabbitmq uses message priority. If you want to use rabbitmq priorities instead, you can do that by setting the priority attribute on the  `AmqpStamp` as introduced in #30913
